feat: use just two SQL calls to back the incidents API

srabraham opened the pull request:
Currently a call to listIncidents (the API that backs the incidents page) results in over 5*n sequential DB lookups, where n is the number of incidents in the relevant event. Given that we build to over 1000 incidents per event, that starts to be quite a lot of queries for each incidents page load. My suspicion is that it'd be much faster to issue just a few big queries instead, and that's what this change does.

On staging, before this change, a listIncidents for 2023 takes about 10 seconds and issues over 6000 DB selects. I'm hopeful that this change will bring that duration down significantly. https://ranger-ims-staging.burningman.org/ims/api/events/2023/incidents/

I've confirmed locally for MySQL and for SQLite that this change doesn't alter the API result for my bunch of test data. The DataStoreIncidentTests pass on this change too, so that's pretty cool.

https://github.com/burningmantech/ranger-ims-server/issues/1324

wsanchez commented on src/ims/store/_db.py:
```diff
@@ -581,6 +585,87 @@ def detachedReportEntries(txn: Transaction) -> Iterable[ReportEntry]:
+        parameters: Parameters = dict(eventID=eventID)
+
+        report_entries = defaultdict[int, list[ReportEntry]](list)
+        txn.execute(self.query.incidents_reportEntries.text, parameters)
```
Quote from the future (via GitHub Time Machine™):
> But @wsanchez, you are such a liar! Right here is `incidents_reportEntries`, which is your fault!

This is true. The `_` indicates a subcategory of some sort. In this case, it's `<tableName>_<queryName>`.

srabraham replied:
Good to know!

wsanchez commented on src/ims/store/mysql/_queries.py:
```diff
@@ -213,6 +213,67 @@
+            i.LOCATION_DESCRIPTION,
+            i.EVENT,
+            (
+                select json_arrayagg(it.NAME)
```
`json_arrayagg` is too cool for me, so I'll just assume this is sane. 🙃

srabraham replied:
haha, yeah, I just learned about it myself. It's relatively new. It takes a bunch of things and turns them into a JSON array (i.e. a string that can be parsed as a JSON array)

wsanchez reviewed (changes requested):
If you could match the code style on the variable names, that would be swell.
Looks good otherwise.

wsanchez commented on src/ims/store/_db.py:
```diff
@@ -674,10 +759,12 @@ async def incidents(self, eventID: str) -> Iterable[Incident]:
-                for number in tuple(self._fetchIncidentNumbers(eventID, txn))
-            )
+            return self._fetchIncidents(eventID, txn)
+            # TODO: remove this if the _fetchIncidents way performs better
+            # return tuple(
+            #     self._fetchIncident(eventID, number, txn)
+            #     for number in tuple(self._fetchIncidentNumbers(eventID, txn))
+            # )
```
Did you confirm this is better?

FWIW, if you add `--log-level=debug` to server startup and look for messages in the namespace `ims.store.mysql.*` (or `ims.store.sqlite.*`), you'll see all of the DB transactions are logged, eg:

```console
2024-10-31T11:19:49-0700 [ims.store.mysql._store.Cursor#debug] EXECUTE: 
	select VERSION from SCHEMA_INFO
	 <- {}
2024-10-31T11:19:49-0700 [ims.store._db.DatabaseManager#debug] No upgrade required for schema version 6
2024-10-31T11:19:49-0700 [ims.store.mysql._store.DataStore#info] Validating data store...
2024-10-31T11:19:50-0700 [ims.store.mysql._store.Cursor#debug] EXECUTE: 
	select AUTHOR, TEXT, CREATED, GENERATED from REPORT_ENTRY
	where
	    ID not in (select REPORT_ENTRY from INCIDENT__REPORT_ENTRY) and
	    ID not in (select REPORT_ENTRY from INCIDENT_REPORT__REPORT_ENTRY)
	 <- {}
```

srabraham replied:
Yeah, I have been watching those logs locally, with a small bunch of incidents and incident entries
I can only run IMS locally and see the logs locally as things are, so my intention is to merge the PR to see the impact on API duration in staging

srabraham commented on src/ims/store/_db.py:
```diff
@@ -581,6 +585,87 @@ def detachedReportEntries(txn: Transaction) -> Iterable[ReportEntry]:
+    ) -> Iterable[Incident]:
+        parameters: Parameters = dict(eventID=eventID)
+
+        report_entries = defaultdict[int, list[ReportEntry]](list)
```
ahh, ok, I'd wondered about that...it's been really weird seeing camelcase in Python. I agree consistency is important. Done

srabraham commented:
I've addressed the review comments. The "change requested" review status is blocking me from merging myself

srabraham commented:
> If you could match the code style on the variable names, that would be swell. Looks good otherwise.

Looks like this is an approval, given that I've addressed all the comments. I'm going to dismiss the "changes requested", which otherwise blocks merging
